### PR TITLE
🐛 ec2/byoip: fix EIP leak when creating machine

### LINKED ISSUE #5038: Prevent leaking EIP when creating machines with BYO IPv4 Pool
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

There are some situations that a machine deployed with BYO IPv4 pool is leaking Elastic IPs:
- 1/ When the instance is created, the flag (`AssociatePublicIpAddress`) to create the instance without public IP must be [set to false][code1] to the primary [network interface][code2], otherwise the instance will be created with an Amazon-provided while the BYO IP reconciliation loop doesn't reach the BYO reconciliation, and the custom EIP is allocated and associated to the instance.
- 2/ The machine reconciliation loop is reaching some race condition or inconsistency from AWS API which is making the controller to create two EIP for each machine created, alternating* between each EIP while the reconciliation finished, leaking one (unused/dissociated) in the instance life cycle. The delete flow is removing both.

*alternating is expected as the algorithm lookup the EIP by `role` trying to optimize/reuse disassociated.

Furthermore, the following failure is happening as the BYO IP reconciliation loop is trying to associate to an non-running machine:

```sh
time="2024-05-08T15:49:33-03:00" level=debug msg="I0508 15:49:33.785472 2878400 recorder.go:104] 
\"Failed to associate Elastic IP for \\\"ec2-i-03de70744825f25c5\\\": InvalidInstanceID: 
The pending instance 'i-03de70744825f25c5' is not in a valid state for this operation.\\n\\tstatus code: 
400, request id: 7582391c-b35e-44b9-8455-e68663d90fed\" logger=\"events\" type=\"Warning\" 
object=[...]\"name\":\"mrb-byoip-32-kbcz9\",\"[...] reason=\"FailedAssociateEIP\""

time="2024-05-08T15:49:33-03:00" level=debug msg="E0508 15:49:33.803742 2878400 controller.go:329] \"Reconciler error\" err=<"

time="2024-05-08T15:49:33-03:00" level=debug msg="\tfailed to reconcile EIP: failed to associate Elastic IP 
\"eipalloc-08faccab2dbb28d4f\" to instance \"i-03de70744825f25c5\": 
InvalidInstanceID: The pending instance 'i-03de70744825f25c5' is not in a valid state for this operation."
```

**What did you expect to happen:**

- Machine is created successfully allocating a single EIP when using BYO IPv4 Pool
- Machine reconciliation loop must wait the instance to leaving the [pending state][ec2-states] before trying to associate EIP, preventing error messages from expected behaviors in the logs

[code1]: https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/1313226a4bf708ca81d8a765aa82a68f90776f4c/pkg/cloud/services/ec2/instances.go#L185-L186
[code2]: https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/1313226a4bf708ca81d8a765aa82a68f90776f4c/pkg/cloud/services/ec2/instances.go#L565-L582
[ec2-states]: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-lifecycle.html

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster-api-provider-aws version: v2.5.2
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`): RHCOS

### PR DESCRIPTION
**Pull request by mtulio**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

The instance creation flow is creating by default EIP to instances even if the BYO IP flow is set. BYO IPv4 creates and associates the EIP to instance after it is created, preventing paying for additional EIP (amazon-provided) when creating the instance when the BYO IPv4 Pool is defined to be used by the machine.

Furthermore, the fix provides additional checks to prevent duplicated EIP in the BYO IP reconciliation loop. The extra checks include running the EIP association many times, while the EIP is already associated, and failures in the log when running the EIP association prematurely - when the instance isn't ready, Eg ec2 in pending state.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #5038

**Special notes for your reviewer**:

**Checklist**:

- [x] squashed commits
- [ ] includes documentation (N/A)
- [x] includes [emojis](https://github.com/kubernetes-sigs/kubebuilder-release-tools?tab=readme-ov-file#kubebuilder-project-versioning)
- [ ] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:

```release-note
fix duplicated/leaked EIP when using BYO IPv4 on Machines.
```

### CODE REVIEW

File: pkg/cloud/services/ec2/eip.go [hunk 1]
```diff
@@ -17,9 +17,40 @@ func getElasticIPRoleName(instanceID string) string {
 
 // ReconcileElasticIPFromPublicPool reconciles the elastic IP from a custom Public IPv4 Pool.
 func (s *Service) ReconcileElasticIPFromPublicPool(pool *infrav1.ElasticIPPool, instance *infrav1.Instance) error {
-	// TODO: check if the instance is in the state allowing EIP association.
-	// Expected instance states: pending or running
-	// https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-lifecycle.html
+	iip := ""
+	if instance.PublicIP != nil {
+		iip = aws.StringValue(instance.PublicIP)
+	}
```
Comment (r4f4) [inline pkg/cloud/services/ec2/eip.go:23]:
```suggestion
	iip := ptr.Deref(instance.PublicIP, "")
```

File: pkg/cloud/services/ec2/eip.go [hunk 2]
```diff
@@ -17,9 +17,40 @@ func getElasticIPRoleName(instanceID string) string {
...
+	s.scope.Debug("Reconciling machine with custom Public IPv4 Pool", "instance-id", instance.ID, "instance-state", instance.State, "instance-public-ip", iip, "publicIpv4PoolID", pool.PublicIpv4Pool)
+
+	// Requeue when the instance is not ready to be associated.
+	if instance.State == infrav1.InstanceStatePending {
```
Comment (r4f4) [inline pkg/cloud/services/ec2/eip.go:27]:
```suggestion
	if instance.State != infrav1.InstanceStateRunning {
```

File: pkg/cloud/services/ec2/instances.go [hunk 3]
```diff
@@ -563,21 +563,13 @@ func (s *Service) runInstance(role string, i *infrav1.Instance) (*infrav1.Instan
...
-
-			if len(i.SecurityGroupIDs) > 0 {
-				input.SecurityGroupIds = aws.StringSlice(i.SecurityGroupIDs)
-			}
+		input.NetworkInterfaces = []*ec2.InstanceNetworkInterfaceSpecification{
+			{
+				DeviceIndex:              aws.Int64(0),
+				SubnetId:                 aws.String(i.SubnetID),
+				Groups:                   aws.StringSlice(i.SecurityGroupIDs),
+				AssociatePublicIpAddress: i.PublicIPOnLaunch,
```
Comment (r4f4) [inline pkg/cloud/services/ec2/instances.go:571]: Ok, I see you already set `PublicIPOnLaunch` to `false`  if EIPPool is set.

File: pkg/cloud/services/ec2/eip.go [hunk 4]
```diff
@@ -17,9 +17,40 @@ func getElasticIPRoleName(instanceID string) string {
...
+		return err
+	}
+
+	// Prevent running association every reconciliation when it is already done.
+	addrs, err := s.netService.GetAddresses(getElasticIPRoleName(instance.ID))
+	if err != nil {
+		s.scope.Error(err, "error checking if addresses exists for Elastic IP Pool to machine", "eip-role", getElasticIPRoleName(instance.ID))
+		return err
+	}
+	if len(addrs.Addresses) > 1 {
```
Comment (r4f4) [inline pkg/cloud/services/ec2/eip.go:39]:
```suggestion
	if len(addrs.Addresses) != 1 {
```

File: pkg/cloud/services/ec2/eip.go [hunk 5]
```diff
@@ -17,9 +17,40 @@ func getElasticIPRoleName(instanceID string) string {
...
+		return fmt.Errorf("unexpected number of EIPs allocated to the role. expected 1, got %d", len(addrs.Addresses))
+	}
+	for _, addr := range addrs.Addresses {
```
Comment (r4f4) [inline pkg/cloud/services/ec2/eip.go:42]: We already error out if there isn't exactly 1 addr, so maybe we do away with the `for` here.

File: pkg/cloud/services/ec2/instances.go (hunk 3 above)
Comment (mtulio) [reply pkg/cloud/services/ec2/instances.go:571]: Yes, preserving the value from machine spec, and forcing to false when EIPPool

https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/1313226a4bf708ca81d8a765aa82a68f90776f4c/pkg/cloud/services/ec2/instances.go#L185-L193

File: pkg/cloud/services/ec2/eip.go (hunk 5 above)
Comment (mtulio) [reply pkg/cloud/services/ec2/eip.go:42]: Thanks! I'll aggregate the checks, do you think we'll have cases where addr.AssociationId!=nil and addr.InstanceId==nil ? It should not happen I think.

File: pkg/cloud/services/ec2/instances.go (hunk 3 above)
Comment (mtulio) [reply pkg/cloud/services/ec2/instances.go:571]: The main problem in this point is leaving the flag unset (original code), the `AssociatePublicIpAddress` seems to assume True in some point making the EIP being created before BYO IP reconciliation loop is reached.

I may have some scenarios triggered by those tests to review:

~~~
--- FAIL: TestCreateInstance (0.01s)
    --- FAIL: TestCreateInstance/when_multiple_subnets_match_filters,_subnets_in_the_cluster_vpc_are_preferred (0.00s)
    --- FAIL: TestCreateInstance/with_a_subnet_outside_the_cluster_vpc (0.00s)
    --- FAIL: TestCreateInstance/with_dedicated_tenancy_cloud-config (0.00s)
    --- FAIL: TestCreateInstance/with_custom_placement_group_cloud-config (0.00s)
    --- FAIL: TestCreateInstance/with_dedicated_tenancy_and_placement_group_ignition (0.00s)
    --- FAIL: TestCreateInstance/with_custom_placement_group_and_partition_number (0.00s)
~~~

Need to review those tests if it is expected or also a bug in the expected return values from API.

File: pkg/cloud/services/ec2/eip.go (hunk 5 above)
Comment (mtulio) [reply pkg/cloud/services/ec2/eip.go:42]: EIPs can be associated to different resources. But in our case (CAPA/Machine) we expect only instanceId

**Comment by mtulio** [comment]
Premature failure.

/test pull-cluster-api-provider-aws-e2e

File: pkg/cloud/services/ec2/instances.go (hunk 3 above)
Comment (mtulio) [reply pkg/cloud/services/ec2/instances.go:571]: changes in `pkg/cloud/services/ec2/instances_test.go` fixes the RunInstanceWithContext api call to adapt to this change, it seems not impact in the use case described in the tests as it will result in the same.

**Comment by mtulio** [comment]
> /test pull-cluster-api-provider-aws-e2e

Okay, previous test failures were flake. The latest run pass. OpenShift e2e BYOIP test is also [passing install][log]: 

~~~
time="2024-06-28T19:16:32Z" level=debug msg="E0628 19:16:32.239704     331 
awsmachine_controller.go:544] \"failed to reconcile BYO Public IPv4\" 
err=\"unable to reconcile Elastic IP Pool to instance \\\"i-05af21b09d3552d0f\\\" with state: pending\""

[...]

time="2024-06-28T19:16:50Z" level=debug msg="I0628 19:16:50.432073     331 eip.go:44] 
\"machine is already associated with an Elastic IP with custom Public IPv4 pool\" 
controller=\"awsmachine\" controllerGroup=\"infrastructure.cluster.x-k8s.io\" 
controllerKind=\"AWSMachine\" AWSMachine=\"openshift-cluster-api-guests/ci-op-f04mmlsn-88881-49rgt-bootstrap\" namespace=\"openshift-cluster-api-guests\" 
name=\"ci-op-f04mmlsn-88881-49rgt-bootstrap\" 
reconcileID=\"825b335b-569b-4a16-891d-c6b7fb5a9db6\" 
machine=\"openshift-cluster-api-guests/ci-op-f04mmlsn-88881-49rgt-bootstrap\" 
cluster=\"openshift-cluster-api-guests/ci-op-f04mmlsn-88881-49rgt\" 
eip=\"eipalloc-0678f0da2c1ecd771\" eip-address=\"157.254.217.22\" 
eip-associationID=\"eipassoc-0ea51b9c68cdfb171\" eip-instance=\"i-05af21b09d3552d0f\""
~~~

This PR is ready for review.  PTAL?
/assign @Ankitasw @dlipovetsky
cc @r4f4 @nrb

/test pull-cluster-api-provider-aws-e2e-eks


[log]: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_installer/8676/pull-ci-openshift-installer-master-altinfra-e2e-aws-ovn-public-ipv4-pool/1806760717966118912/artifacts/e2e-aws-ovn-public-ipv4-pool/ipi-install-install/artifacts/.openshift_install-1719604107.log

File: controllers/awsmachine_controller.go [hunk 6]
```diff
@@ -541,7 +541,7 @@ func (r *AWSMachineReconciler) reconcileNormal(_ context.Context, machineScope *
 	// a BYOIP without duplication.
 	if pool := machineScope.GetElasticIPPool(); pool != nil {
 		if err := ec2svc.ReconcileElasticIPFromPublicPool(pool, instance); err != nil {
-			machineScope.Error(err, "failed to associate elastic IP address")
+			machineScope.Error(err, "failed to reconcile BYO Public IPv4")
```
Comment (r4f4) [inline controllers/awsmachine_controller.go:544]: What do you think about what I did with registering the instance with the LB: https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/5040/files#diff-0b559bbd149f0e6d54d789235423f66fa8906fdc6ee9c99b9e85db912912011eR622-R629. Just returning an error because the instance is not yet running might confuse users looking at the logs.

Comment (mtulio) [reply controllers/awsmachine_controller.go:544]: @r4f4 that's a great idea, specially now it is generating more warnings with this approach, instead of leak and failures. But non error/warn messages for expected states would be much better.

**Comment by mtulio** [comment]
e2e EKS test is passing.

I am holding this PR to address the @r4f4 's [feedback](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/5039#discussion_r1666064699) to decrease the warnings in the code, silent re-queuing expected states.

/hold

File: controllers/awsmachine_controller.go (hunk 6 above)
Comment (mtulio) [reply controllers/awsmachine_controller.go]: @r4f4 resolving this thread, feedback addressed in the most recent changes. This [comment](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/5039#issuecomment-2315984987) has more details.

Would you mind taking a look?

**Comment by r4f4** [comment]
LGTM

**Comment by mtulio** [comment]
> /test pull-cluster-api-provider-aws-e2e

still hitting infra issues/timeouts:

/test pull-cluster-api-provider-aws-e2e
/test pull-cluster-api-provider-aws-e2e-eks

**Comment by nrb** [comment]
/retest

Will re-review today.

**Comment by mtulio** [comment]
Still seeing e2e issues that seems to be unrelated. Thanks @nrb for re-triggering. Looks like e2e-eks is [passing][1829164109933842432].

I will trigger the regular e2e again:

/test pull-cluster-api-provider-aws-e2e

Downstream jobs are triggered too: https://github.com/openshift/installer/pull/8676#issuecomment-2318667965



[1829164109933842432]: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/5039/pull-cluster-api-provider-aws-e2e-eks/1829164109933842432

**Comment by mtulio** [comment]
Hi @nrb , would you mind taking a review in this bug, please?

As you see in the last comments, we are struggling to run the job `pull-cluster-api-provider-aws-e2e`, but the `e2e-eks`, and `ele-conformance` are passing (not sure under the hood if it's all touching this change, but considering this change is changing Machine creation flow, it think it should).

Furthermore, in downstream/OpenShift we are running several presubmit jobs (Public IPv4 pool is default over aws jobs) across the PR https://github.com/openshift/installer/pull/8676 (vendoring this PR). I also introduced  a new presubmit (https://github.com/openshift/release/pull/56114) to enforce to disable the pool in CAPA provisioning to test the non-pool flow and it is all passing: https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_installer/8676/pull-ci-openshift-installer-master-e2e-aws-ovn-public-ipv4-pool-disabled/1829621227333881856

I will trigger again, but also open to hear from you if you could share your tougths of another job to validate it, or those presented is ok.

Looking forward to hear from you, thanks!

cc @r4f4 @patrickdillon

/test pull-cluster-api-provider-aws-e2e

**Comment by mtulio** [comment]
tl;dr: looks like `pull-cluster-api-provider-aws-e2e` is failing for unrelated change in this PR, it is trying to create a cluster with an AMI that does not exist in the test account.

---

After some investigation with @nrb, we are seeing the job `e2e` always failing the test ` [It] [unmanaged] [Cluster API Framework] Clusterctl Upgrade Spec [from latest v1beta1 release to v1beta2] Should create a management cluster and then upgrade all the providers` .

We are seeing in the Control Plane spec that the AMI for kube 1.25 isn't available:
```
  - lastTransitionTime: "2024-09-02T15:00:37Z"
    message: 'failed to create AWSMachine instance: failed to find ami: found no AMIs
      with the name: "capa-ami-ubuntu-18.04-?1.25.0-*"'
```

Looks like the CAPI e2e[1] is setting the following variable:

https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/abe918cfe1038c714792860b23632e38c7e0438f/test/e2e/data/e2e_conf.yaml#L207-L209

in the test spec:
https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/abe918cfe1038c714792860b23632e38c7e0438f/test/e2e/suites/unmanaged/unmanaged_CAPI_test.go#L112-L129

Causing the failures when looking up for an AMI that does not exists in the test account. (maybe had been pruned or 1.25 is not supported and dont need anymore?)

[1] https://github.com/kubernetes-sigs/cluster-api/blob/main/test/e2e/clusterctl_upgrade.go#L237C33-L237C58

**Comment by mtulio** [comment]
For reviewers: this PR is general ready for review. Following my last comment, the failure is unrelated with this PR.

File: pkg/cloud/services/ec2/eip.go [hunk 7]
```diff
@@ -16,17 +17,46 @@ func getElasticIPRoleName(instanceID string) string {
...
+
+	// Prevent running association every reconciliation when it is already done.
+	addrs, err := s.netService.GetAddresses(getElasticIPRoleName(instance.ID))
+	if err != nil {
+		s.scope.Error(err, "error checking if addresses exists for Elastic IP Pool to machine", "eip-role", getElasticIPRoleName(instance.ID))
+		return shouldRequeue, err
+	}
+	if len(addrs.Addresses) > 0 {
+		if len(addrs.Addresses) != 1 {
+			return shouldRequeue, fmt.Errorf("unexpected number of EIPs allocated to the role. expected 1, got %d", len(addrs.Addresses))
```
Comment (nrb) [inline pkg/cloud/services/ec2/eip.go:42]: In what case would this happen? A user manually attaching an EIP to the role/instance?

File: pkg/cloud/services/ec2/eip.go [hunk 8]
```diff
@@ -16,17 +17,46 @@ func getElasticIPRoleName(instanceID string) string {
...
+		if addr.AssociationId != nil && addr.InstanceId != nil && *addr.InstanceId == instance.ID {
+			s.scope.Debug("Machine is already associated with an Elastic IP with custom Public IPv4 pool", "eip", addr.AllocationId, "eip-address", addr.PublicIp, "eip-associationID", addr.AssociationId, "eip-instance", addr.InstanceId)
+			return shouldRequeue, nil
+		}
+	}
+
+	// Associate EIP.
 	if err := s.getAndAssociateAddressesToInstance(pool, getElasticIPRoleName(instance.ID), instance.ID); err != nil {
-		return fmt.Errorf("failed to reconcile Elastic IP: %w", err)
+		return shouldRequeue, fmt.Errorf("failed to reconcile Elastic IP: %w", err)
```
Comment (nrb) [inline pkg/cloud/services/ec2/eip.go:54]: `shouldRequeue` is `false` here, correct? So the only time we ever requeue is if the instance is not yet running, and all other states are terminal?

**Comment by nrb** [comment]
/override pull-cluster-api-provider-aws-e2e

This test is failing due to something unrelated right now.

File: pkg/cloud/services/ec2/eip.go (hunk 7 above)
Comment (mtulio) [reply pkg/cloud/services/ec2/eip.go:42]: This is the main reason of this PR. The controller was allocating more than one EIP in the reconciliation loop. Another possibility I can see would be manually (or controller issue) allocating an EIP with the same role (`eip-<instanceID>`)

File: pkg/cloud/services/ec2/eip.go (hunk 8 above)
Comment (mtulio) [reply pkg/cloud/services/ec2/eip.go]: fixed

File: pkg/cloud/services/ec2/eip.go [hunk 9]
```diff
@@ -16,17 +17,51 @@ func getElasticIPRoleName(instanceID string) string {
...
+		addr := addrs.Addresses[0]
+		// address is already associated.
+		if addr.AssociationId != nil && addr.InstanceId != nil && *addr.InstanceId == instance.ID {
+			s.scope.Debug("Machine is already associated with an Elastic IP with custom Public IPv4 pool", "eip", addr.AllocationId, "eip-address", addr.PublicIp, "eip-associationID", addr.AssociationId, "eip-instance", addr.InstanceId)
+			return shouldRequeue, nil
+		}
+	}
+
+	// Get existing, or allocate an EIP, then Associate to the machine.
+	// Should requeue if any error is returned in the process.
```
Comment (nrb) [inline pkg/cloud/services/ec2/eip.go:57]: Could we move this above L36 (before `shouldRequeue` is `false`), or must it come after the address length check and checking that the machine already has an elastic IP?

Just thinking from an organizational stand point, if we could group all the requeue cases together that would be helpful, but understandable if not.

**Comment by nrb** [comment]
I don't think the failure is related to this PR, nor the previous image issues we were seeing; the `AWSMachine` resources get to a `Ready` state and don't have errors themselves.

/retest

File: pkg/cloud/services/ec2/eip.go (hunk 9 above)
Comment (mtulio) [reply pkg/cloud/services/ec2/eip.go:57]: Thanks for suggestion. Reviewing here, the order is important to be idempotent. The lines [L38-L54](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/4626a6afbb6bb7f212568f24981aa9bb79929f07/pkg/cloud/services/ec2/eip.go#L38-L54)
should check if there is already an address and is not assigned before allocating a new one L58 (`getAndAssociateAddressesToInstance()`).

I can move the logic of those lines to `getAndAssociateAddressesToInstance()`, but it will need to be validated in either way.

**Comment by mtulio** [comment]
> @mtulio: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> pull-cluster-api-provider-aws-e2e	[4626a6a](https://github.com/kubernetes-sigs/cluster-api-provider-aws/commit/4626a6afbb6bb7f212568f24981aa9bb79929f07)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/5039/pull-cluster-api-provider-aws-e2e/1833203837066809344)	false	`/test pull-cluster-api-provider-aws-e2e`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes-sigs&repo=cluster-api-provider-aws&pr=5039). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Amtulio). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues?q=is:issue+is:open) when you hit one in your PR.

I am seeing a lot CloudFormation to provision required environment, I can't see if could be related.

/retest

**Comment by mtulio** [comment]
@nrb e2e passing now! 🎉
